harness/determined/profiler.py

```diff
@@ -18,19 +18,96 @@
 LOG_NAMESPACE = "determined-profiler"
 
-try:
-    import pynvml
 
-    pynvml.nvmlInit()
-    SHOULD_PROFILE_GPUS = True
-except ModuleNotFoundError:
-    logging.info(f"{LOG_NAMESPACE}: pynvml not found. Not collecting GPU metrics")
-    SHOULD_PROFILE_GPUS = False
-except pynvml.NVMLError_LibraryNotFound:
-    logging.info(f"{LOG_NAMESPACE}: pynvml LibraryNotFound error. Not collecting GPU metrics")
-    SHOULD_PROFILE_GPUS = False
-except Exception as e:
-    raise RuntimeError(f"Could not set up pynvml, but it failed with an unexpected error: {e}")
+class PynvmlWrapperError(Exception):
+    pass
+
+
+class PynvmlWrapper:
+    def __init__(self) -> None:
+        self._pynvml = None  # type: Optional[Any]
+        self._device_count = None  # type: Optional[int]
+        self._index_to_uuid_map = {}  # type: Dict[int, str]
+        try:
+            import pynvml
+
+            pynvml.nvmlInit()
+            try:
+                num_gpus = pynvml.nvmlDeviceGetCount()
+                for i in range(num_gpus):
+                    handle = pynvml.nvmlDeviceGetHandleByIndex(i)
+                    pynvml.nvmlDeviceGetUUID(handle)
+                    pynvml.nvmlDeviceGetMemoryInfo(handle)
+                    pynvml.nvmlDeviceGetUtilizationRates(handle)
+                self._pynvml = pynvml
+            except Exception as e:
+                logging.warning(
+                    f"{LOG_NAMESPACE}: pynvml is functional, but failed to pass functionality "
+                    f"test due to exception. Not collecting GPU metrics. Exception details: {e}"
+                )
+        except ModuleNotFoundError:
+            logging.info(f"{LOG_NAMESPACE}: pynvml not found. Not collecting GPU metrics")
+        except pynvml.NVMLError_LibraryNotFound:
+            logging.info(
+                f"{LOG_NAMESPACE}: pynvml LibraryNotFound error. Not collecting GPU metrics"
+            )
+        except Exception as e:
+            logging.error(
+                f"{LOG_NAMESPACE}: unexpected error while trying to set up pynvml. Not "
+                f"collecting GPU metrics. Please report this error to "
+                f"https://github.com/determined-ai/determined as it should not be "
+                f"encountered by users. Error details: {e}"
+            )
+
+    @property
+    def pynvml_is_available(self) -> bool:
+        return self._pynvml is not None
+
+    def _safety_check(self) -> None:
+        """Before calling any pynvml operations, raise an error if pynvml is not available"""
+        if self._pynvml is None:
+            raise PynvmlWrapperError(
+                "Tried to call a pynvml operation but pynvml is either unavailable or not "
+                "functional. Code should check pynvml_is_working before calling any operations."
+            )
+
+    def nvml_get_device_count(self) -> int:
+        self._safety_check()
+        self._pynvml = cast(Any, self._pynvml)
+
+        if self._device_count is None:
```

Comment:
could we just set this in init and make this a prop?

Reply:
fixed